### PR TITLE
Add AIP for V4 AL Service Provider Proposal

### PR DESCRIPTION
**Pull request by CheyenneAtapour**
https://governance.aave.com/t/temp-check-service-provider-proposal/17866

### CODE REVIEW

**Review by miguelmtzinf — CHANGES_REQUESTED** [review]

File: src/20240614_AaveV3Ethereum_V4ALServiceProviderProposal/AaveV3Ethereum_V4ALServiceProviderProposal_20240614.sol [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+  // 3 million GHO upfront
+  uint256 public constant GHO_UPFRONT_AMOUNT = 3_000_000 ether;
+
+  // 9 million GHO streamed over the year
+  uint256 public constant GHO_STREAM_AMOUNT = 9_000_000 ether;
+  uint256 public constant GHO_STREAM_DURATION = 365 days;
+
+  uint256 public constant ACTUAL_STREAM =
+    (GHO_STREAM_AMOUNT / GHO_STREAM_DURATION) * GHO_STREAM_DURATION;
```
[the remark this answers is not in the document]
Comment (CheyenneAtapour) [reply src/20240614_AaveV3Ethereum_V4ALServiceProviderProposal/AaveV3Ethereum_V4ALServiceProviderProposal_20240614.sol:25]: It ensures there are no rounding errors with the stream by ensuring the dollar amount is divisible by seconds. I saw the same was done here: https://github.com/bgd-labs/aave-proposals-v3/pull/331/files#diff-19fff750e1e69bf557e3608c28b57945f6b0cfdc33b99b4988c4494a2de10ce6R23

File: src/20240614_AaveV3Ethereum_V4ALServiceProviderProposal/AaveV3Ethereum_V4ALServiceProviderProposal_20240614.t.sol [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+      address(AaveV3Ethereum.COLLECTOR),
+      3_000_000 ether
+    );
+
+    uint256 CollectorV3GHOBalanceBefore = IERC20(AaveV3EthereumAssets.GHO_UNDERLYING).balanceOf(
+      address(AaveV3Ethereum.COLLECTOR)
+    );
+
+    executePayload(vm, address(proposal));
+
```
[the remark this answers is not in the document]
Comment (CheyenneAtapour) [reply src/20240614_AaveV3Ethereum_V4ALServiceProviderProposal/AaveV3Ethereum_V4ALServiceProviderProposal_20240614.t.sol:129]: done

File: src/20240614_AaveV3Ethereum_V4ALServiceProviderProposal/AaveV3Ethereum_V4ALServiceProviderProposal_20240614.sol [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+/**
+ * @title V4 AL Service Provider Proposal
+ * @author Aave Labs
+ * - Snapshot: https://snapshot.org/#/aave.eth/proposal/0xbf901f4be94a4661dce8217b3b037a8607ea8953cbe32e7dbde6a882819d64b3
+ * - Discussion: https://governance.aave.com/t/temp-check-service-provider-proposal/17866
+ */
+contract AaveV3Ethereum_V4ALServiceProviderProposal_20240614 is IProposalGenericExecutor {
+  // TODO: Find the appropriate recipient address for AAVE Labs
+  address public constant AAVE_LABS = 0xac140648435d03f784879cd789130F22Ef588Fcd;
```
[the remark this answers is not in the document]
Comment (CheyenneAtapour) [reply src/20240614_AaveV3Ethereum_V4ALServiceProviderProposal/AaveV3Ethereum_V4ALServiceProviderProposal_20240614.sol]: done

File: src/20240614_AaveV3Ethereum_V4ALServiceProviderProposal/AaveV3Ethereum_V4ALServiceProviderProposal_20240614.sol [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+pragma solidity ^0.8.0;
+
+import {IProposalGenericExecutor} from 'aave-helpers/interfaces/IProposalGenericExecutor.sol';
+import {AaveV3Ethereum, AaveV3EthereumAssets} from 'aave-address-book/AaveV3Ethereum.sol';
+
+/**
+ * @title V4 AL Service Provider Proposal
+ * @author Aave Labs
+ * - Snapshot: https://snapshot.org/#/aave.eth/proposal/0xbf901f4be94a4661dce8217b3b037a8607ea8953cbe32e7dbde6a882819d64b3
+ * - Discussion: https://governance.aave.com/t/temp-check-service-provider-proposal/17866
```
Comment (miguelmtzinf) [inline src/20240614_AaveV3Ethereum_V4ALServiceProviderProposal/AaveV3Ethereum_V4ALServiceProviderProposal_20240614.sol:11]: Update these two:
https://snapshot.org/#/aave.eth/proposal/0x70dfd865b78c4c391e2b0729b907d152e6e8a0da683416d617d8f84782036349
https://governance.aave.com/t/arfc-al-service-provider-proposal/17974

File: src/20240614_AaveV3Ethereum_V4ALServiceProviderProposal/AaveV3Ethereum_V4ALServiceProviderProposal_20240614.t.sol [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+    vm.stopPrank();
+  }
+
+  // Test giving V2 Collector more funds and full withdrawing
+  function testProposalExecutionPrankFunds() public {
+    uint256 nextCollectorStreamID = AaveV3Ethereum.COLLECTOR.getNextStreamId();
+    uint256 ALGHOBalanceBefore = IERC20(AaveV3EthereumAssets.GHO_UNDERLYING).balanceOf(AAVE_LABS);
+
+    // Giving the Collector enough funds to cover the stream
+    vm.prank(0x1a88Df1cFe15Af22B3c4c783D4e6F7F9e0C1885d);
```
Comment (miguelmtzinf) [inline src/20240614_AaveV3Ethereum_V4ALServiceProviderProposal/AaveV3Ethereum_V4ALServiceProviderProposal_20240614.t.sol:118]: can we just use `deal`?

File: src/20240614_AaveV3Ethereum_V4ALServiceProviderProposal/config.ts [hunk 6]
```diff
@@ -0,0 +1,14 @@
+import {ConfigFile} from '../../generator/types';
+export const config: ConfigFile = {
+  rootOptions: {
+    title: 'V4 AL Service Provider Proposal',
+    author: 'Aave Labs',
+    discussion: 'https://governance.aave.com/t/temp-check-service-provider-proposal/17866',
+    snapshot:
```
Comment (miguelmtzinf) [inline src/20240614_AaveV3Ethereum_V4ALServiceProviderProposal/config.ts:7]: update

File: src/20240614_AaveV3Ethereum_V4ALServiceProviderProposal/AaveV3Ethereum_V4ALServiceProviderProposal_20240614.sol (hunk 4 above)
Comment (CheyenneAtapour) [reply src/20240614_AaveV3Ethereum_V4ALServiceProviderProposal/AaveV3Ethereum_V4ALServiceProviderProposal_20240614.sol]: done

File: src/20240614_AaveV3Ethereum_V4ALServiceProviderProposal/AaveV3Ethereum_V4ALServiceProviderProposal_20240614.t.sol (hunk 5 above)
Comment (CheyenneAtapour) [reply src/20240614_AaveV3Ethereum_V4ALServiceProviderProposal/AaveV3Ethereum_V4ALServiceProviderProposal_20240614.t.sol]: done

File: src/20240614_AaveV3Ethereum_V4ALServiceProviderProposal/config.ts (hunk 6 above)
Comment (CheyenneAtapour) [reply src/20240614_AaveV3Ethereum_V4ALServiceProviderProposal/config.ts]: done

**Review by sakulstra — DISMISSED** [review]

File: src/20240614_AaveV3Ethereum_V4ALServiceProviderProposal/V4ALServiceProviderProposal.md [hunk 7]
```diff
@@ -0,0 +1,112 @@
+---
+title: "[ARFC] AL Service Provider Proposal"
```
Comment (sendra) [inline src/20240614_AaveV3Ethereum_V4ALServiceProviderProposal/V4ALServiceProviderProposal.md:2]: No need to put ARFC in the title

File: src/20240614_AaveV3Ethereum_V4ALServiceProviderProposal/V4ALServiceProviderProposal.md [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+author: "Aave Labs"
+discussions: "https://governance.aave.com/t/arfc-al-service-provider-proposal/17974"
+snapshot: "https://snapshot.org/#/aave.eth/proposal/0x70dfd865b78c4c391e2b0729b907d152e6e8a0da683416d617d8f84782036349"
+---
+
+## Summary
+
+This ARFC proposes a one-year technical contributor engagement for the Aave Labs to act as a service provider for development of Aave V4 and a new visual identity.
```
Comment (sendra) [inline src/20240614_AaveV3Ethereum_V4ALServiceProviderProposal/V4ALServiceProviderProposal.md:10]: This is the AIP not ARFC

File: src/20240614_AaveV3Ethereum_V4ALServiceProviderProposal/V4ALServiceProviderProposal.md [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+## Principles
+
+- Aave Labs shall strictly adhere to the scope of its designated role as one of the service providers working as a technical contributor, ensuring all decisions and actions remain within our areas of the engagement.
+- Aave Labs shall prioritize the security and robustness of its technical contributions.
+- Any production-ready code written within the scope of the proposal will be moved to the Aave DAO’s GitHub repository.
+- Aave Labs shall engage with the DAO like all other contributors within the established working relationships and procedures of the Aave DAO. For example, decisions concerning risk configurations, DAO treasury, code audits, and other matters not covered by this proposal will need to be separately approved by the Aave DAO via TEMP CHECKs and Snapshots.
+
+## Copyright
+
+The text of this ARFC is released under the [CC0 license](https://creativecommons.org/publicdomain/zero/1.0/). The visuals and the New Visual identity are subject to and governed by the license specified in the approved governance proposal by the Aave DAO (to be included in the ARFC at a later date).
```
Comment (sendra) [inline src/20240614_AaveV3Ethereum_V4ALServiceProviderProposal/V4ALServiceProviderProposal.md:112]: AIP

File: src/20240614_AaveV3Ethereum_V4ALServiceProviderProposal/V4ALServiceProviderProposal.md [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+## Principles
+
+- Aave Labs shall strictly adhere to the scope of its designated role as one of the service providers working as a technical contributor, ensuring all decisions and actions remain within our areas of the engagement.
+- Aave Labs shall prioritize the security and robustness of its technical contributions.
+- Any production-ready code written within the scope of the proposal will be moved to the Aave DAO’s GitHub repository.
+- Aave Labs shall engage with the DAO like all other contributors within the established working relationships and procedures of the Aave DAO. For example, decisions concerning risk configurations, DAO treasury, code audits, and other matters not covered by this proposal will need to be separately approved by the Aave DAO via TEMP CHECKs and Snapshots.
+
+## Copyright
+
+The text of this AIP is released under the [CC0 license](https://creativecommons.org/publicdomain/zero/1.0/). The visuals and the New Visual identity are subject to and governed by the license specified in the approved governance proposal by the Aave DAO (to be included in the ARFC at a later date).
```
Comment (sendra) [inline src/20240614_AaveV3Ethereum_V4ALServiceProviderProposal/V4ALServiceProviderProposal.md:112]:
```suggestion
The text of this AIP is released under the [CC0 license](https://creativecommons.org/publicdomain/zero/1.0/). The visuals and the New Visual identity are subject to and governed by the license specified in the approved governance proposal by the Aave DAO.
```

**Review by sendra — APPROVED** [review]

**Pull request merged**